src/behavior-considered-undefined.md

```diff
@@ -155,6 +127,46 @@ entire range, so it is important that the length metadata is never too large. In
 particular, the dynamic size of a Rust value (as determined by `size_of_val`)
 must never exceed `isize::MAX`.
 
+### Invalid values
+[invalid-values]: #invalid-values
+
+The Rust compiler assumes that all values produced during program execution are
+"valid", and producing an invalid value is hence immediate UB.
+
+Whether a value is valid depends on the type:
+* A [`bool`] value must be `false` (`0`) or `true` (`1`).
+* A `fn` pointer value must be non-null.
+* A `char` value must not be a surrogate (i.e., must not be in the range `0xD800..=0xDFFF`) and must be below `char::MAX`.
+* A `!` value must never exist.
+* An integer (`i*`/`u*`), floating point value (`f*`), or raw pointer must be
+  initialized, i.e., must not be obtained from [uninitialized memory][undef].
+* A `str` value is treated like `[u8]`, i.e. it must be initialized.
+* An `enum` must have a valid discriminant, and all fields of the variant indicated by that discriminant must be valid at their respective type.
+* A `struct`, tuple, and array requires all fields/elements to be valid at their respective type.
+* For a `union`, the exact validity requirements are not decided yet. The following is certain:
+  * If the `union` has a zero-sized field, then all values are valid.
+  * If a value is valid for a particular `union` field, then it is valid for the union.
```

Comment:
This isn't exhaustive on what is possible with safe code. For example, it implicitly excludes the value `[0,0]` for the union:
```rust
#[repr(C)]
pub union Foo{
    x: (NonZero<u8>, u8),
    y: (u8, NonZero<u8>),
}
```
which can be created using the purely safe code:
```rust
let mut x = Foo{x: NonZero::new(1).unwrap(), y: 0};
x.y.0 = 0;
let x = x;
``` 

I'm not sure how to modify this to support that, but I'd expect we shouldn't exclude values that are definitely valid (because they are constructable in safe code) if we can avoid it.

Reply:
I know that this doesn't cover everything one can do in safe code. It doesn't claim to do that.

Reply:
If we want this to be able to capture everything safe code can do, we could try something like "if a value satisfies the validity invariant, then picking any (nested) field of an enum and writing a value valid at that field's type to that field yields a resulting union value that is still valid", or so. I think most readers would probably be more confused by this than it helps them...

This isn't meant to summarize all we know about unions, just some basic facts. Currently the page states nothing about unions. So another alternative would be to just say "is not decided yet" or "is subject to active discussion" and link to the [UCG issue](https://github.com/rust-lang/unsafe-code-guidelines/issues/438).

Reply:
Maybe it could be put into a sub bullet or a note.

I just don't think it's a good idea to omit values that are possible to create in safe code.